core/src/main/java/feast/core/service/JobService.java

```diff
@@ -0,0 +1,223 @@
+
+    if (filter.getId() != "") {
+      // get by id: no more filters required: found job
+      Optional<Job> job = this.jobRepository.findById(filter.getId());
+      if (job.isPresent()) {
+        matchingJobIds.add(filter.getId());
+      }
+
+    } else {
+      // multiple filters can apply together in an 'and' operation
+      if (filter.getStoreName() != "") {
+        // find jobs by name
+        List<Job> jobs = this.jobRepository.findByStoreName(filter.getStoreName());
+        List<String> jobIds =
+            jobs.stream()
+                .map(
+                    job -> {
+                      return job.getId();
+                    })
+                .collect(Collectors.toList());
+
+        matchingJobIds = this.mergeResults(matchingJobIds, jobIds);
+      }
+      if (filter.hasFeatureSetReference()) {
+        // find a matching featureset for reference
+        FeatureSetReference fsReference = filter.getFeatureSetReference();
+        List<FeatureSet> matchFeatureSets = this.findFeatureSets(fsReference);
+        Collection<Job> jobs = this.jobRepository.findByFeatureSetIn(matchFeatureSets);
+
+        List<String> jobIds =
+            jobs.stream()
+                .map(
+                    job -> {
+                      return job.getId();
+                    })
+                .collect(Collectors.toList());
+        matchingJobIds = this.mergeResults(matchingJobIds, jobIds);
+      }
+    }
+
+    // convert matching job models to ingestion job protos
+    List<IngestionJobProto.IngestionJob> ingestJobs = new ArrayList<>();
+    for (String jobId : matchingJobIds) {
+      Job job = this.jobRepository.findById(jobId).get();
+      ingestJobs.add(job.toIngestionProto());
+    }
+
+    // pack jobs into response
+    return ListIngestionJobsResponse.newBuilder().addAllJobs(ingestJobs).build();
+  }
+
+  // TODO: restart ingestion job
+
+  /**
+   * Stops (Aborts) the ingestion job matching the given request. Does nothing if the target job to
+   * be stopped is already stopped or stopping
+   *
+   * @param request stop ingestion job request specifying which job to stop
+   * @throws NoSuchElementException when stop job request requests to stop a nonexistent job.
+   * @throws UnsupportedOperationException when job to be stopped is in an unknown status
+   * @throws InvalidProtocolBufferException on error when constructing response protobuf
+   */
+  public StopIngestionJobResponse stopJob(StopIngestionJobRequest request)
+      throws InvalidProtocolBufferException {
+    // check job exists
+    Optional<Job> getJob = this.jobRepository.findById(request.getId());
+    if (getJob.isEmpty()) {
+      throw new NoSuchElementException(
+          "Attempted to stop nonexistent job with id: " + getJob.get().getId());
+    }
+
+    // check job status is valid for stopping
+    Job job = getJob.get();
+    JobStatus status = job.getStatus();
+    if (status.equals(JobStatus.ABORTED)
```

Comment:
You can use `JobStatus.TERMINAL_STATE` for this check

Reply:
How should calling `stopJob()` in one of the transitional states (ie`ABORTING`, `SUSPENDING` be handled? `JobStatus.TERMINAL_STATE` only the final states.

Reply:
I updated JobService use `JobStatus.TERMINAL_STATE` to check for final states. I added `JobStatus.TRANSITIONAL_STATES` to track transitional states (ie `PENDING`, `ABORTING`, `SUSPENDING`)